### DISCUSSION TITLE
Why the mono_fdhandle_insert doesn't release the mutex lock after getting an error

### DISCUSSION THREAD

**Discussion by duduita** [dotnet/runtime#94628] [Q&A]
While most of the methods in `fdhandle.c` release the mutex lock after getting an error, such as `mono_fdhandle_try_insert`, `mono_fdhandle_lookup_and_ref`, `mono_fdhandle_close`, etc... the `mono_fdhandle_insert` doesn't do it, so if for some reason I got a duplicated socket error, the `mono_fdhandle_insert` will fail calling the `g_error` without releasing the mutex lock. Consequently, after that I'm getting the error `mono_os_mutex_trylock: pthread_mutex_trylock failed with "Error code '142'"`, since it seems that other threads cannot lock the mutex since it is already locked.

To summarize, I would like to understand if not releasing the mutex lock after getting an error in the `mono_fdhandle_insert` is intentional, in that case, why we cannot release the mutex as we do in the other functions. Since it's a core method, I guess that it should be made intentionally @luhenry.

**Comment by Xyncgas** [comment]
Maybe because after double failure no one is permitted to enter a crime scene

**Comment by duduita** [comment]
But, why do we use the `mono_fdhandle_try_insert`, which basically does the same as the `mono_fdhandle_insert`, but just releases the mutex lock in the case of failure, as can be seen on the `w32file-unix.c`. My naive question is why we cannot do the same on the `w32socket.c`.
